### PR TITLE
Add gas specific integration test cases

### PR DESCRIPTION
**Pull request by simonlechner**
This PR introduces gas specific integration test cases. It extends the integration test scenario with an expected opera error, since the opera reference implementation handles insufficient gas with errors.

### CODE REVIEW

File: go/integration_test/processor/scenario.go
Comment (facuMH) [inline go/integration_test/processor/scenario.go:32]: this should an implementation agnostic name, maybe `ProcessorError` or `TxError`, but not `OperaError`, wdyt?

File: go/integration_test/processor/processor_gas_test.go
Comment (facuMH) [inline go/integration_test/processor/processor_gas_test.go:111]: why not use `PUSH0` since it should be implemented in all interpreters ?

File: go/integration_test/processor/processor_gas_test.go
Comment (facuMH) [inline go/integration_test/processor/processor_gas_test.go:185]: this error is used multiple times, maybe it's worth predeclaring as a variable or const.

**Review by facuMH — APPROVED** [review]
Looks good, just a couple of style suggestions. Also, the gas limit  formulas are repeated (or patially repeated) for multiple test cases. do you think it would make sense (or possible) to give them significant variable names and reuse them throughout the tests ? this way each formula would only be declared once, less error prone. But I can also see this as a personal style preference, wdyt  ?

File: go/integration_test/processor/processor_gas_test.go
Comment (HerbertJordan) [inline go/integration_test/processor/processor_gas_test.go:86]: We once started discussing this but did not get to a conclusion; I would assume that in such a case the GasLimit might have to be returned, since this corresponds to the actual gas cost, right? After all, I would assume that the gas bought by the user at the begin of the transaction is not returned.

This would also make this case handled identically to a case where one is running out of gas during a contract execution.

File: go/integration_test/processor/processor_gas_test.go
Comment (HerbertJordan) [inline go/integration_test/processor/processor_gas_test.go:104]: nit: "codePrice" => "executionGas" or "executionGasCosts"

File: go/integration_test/processor/processor_gas_test.go
Comment (HerbertJordan) [reply go/integration_test/processor/processor_gas_test.go:111]: I would prefer the `PUSH1` version as it is here, since it would make those tests independent of the revision.

File: go/integration_test/processor/processor_gas_test.go
Comment (HerbertJordan) [inline go/integration_test/processor/processor_gas_test.go:250]: Maybe a helper function producing test cases "<name>Exact", "<name>Sufficient", and "<name>Insufficient" for a given scenario would be a nice to have. Wdyt?

File: go/integration_test/processor/processor_gas_test.go
Comment (HerbertJordan) [inline go/integration_test/processor/processor_gas_test.go:311]: As argued above, I find it surprising that the reported `GasUsed` here is 0 instead of `GasLimit`. We need to check this.

File: go/integration_test/processor/processor_gas_test.go
Comment (HerbertJordan) [inline go/integration_test/processor/processor_gas_test.go:37]: nit: it is a bit odd to use constants from a concrete implementation to define implementation agnostic tests. It might be nice to have those constants defined at some implementation-independent place, but I am not sure what a good place might be. So I guess we leave this for a potential future cleanup pass ...

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
We should re-check the reported `GasUsed` values for failed operations. This might simply also be wrong in the Opera implementation. Reporting values of 0 seem to be off.

File: go/integration_test/processor/processor_gas_test.go
Comment (facuMH) [reply go/integration_test/processor/processor_gas_test.go:37]: I like the idea of implementation agnostic tests, I would say that if we want these values to also be called in the implementations, then a nice place would be on the same level as the processor interface definition. But if we just want these values to be test constants, then they could be defined in `go/integration_tests/processor/gas_values.go` . wdyt ?

**Review by LuisPH3 — COMMENTED** [review]
I looks good to me, please don't wait for my approval. The missing details may be a little too domain specific for my current understanding.

File: go/processor/floria/processor.go
Comment (simonlechner) [reply go/processor/floria/processor.go:20]: This has been taken from the reference implementation.

File: go/integration_test/processor/scenario.go
Comment (simonlechner) [reply go/integration_test/processor/scenario.go:32]: Our implementation will not return error for the tested cases, that's why I specifically named the field `OperaError`.

File: go/integration_test/processor/processor_gas_test.go
Comment (simonlechner) [reply go/integration_test/processor/processor_gas_test.go:185]: The code has been refactored to only use this error in one place.

File: go/integration_test/processor/processor_gas_test.go
Comment (simonlechner) [reply go/integration_test/processor/processor_gas_test.go:37]: I will keep this as a cleanup task once the implementation is fixed. I think all these constants should be in the same place eventually.

File: go/integration_test/processor/processor_gas_test.go
Comment (simonlechner) [reply go/integration_test/processor/processor_gas_test.go:86]: I also thought this is very surprising in the current reference implementation, but I thought it has to be correct because the whole history passed. I changed the reference implementation on this branch, it still has been confirmed to be compatible with the history.

File: go/integration_test/processor/scenario.go
Comment (simonlechner) [reply go/integration_test/processor/scenario.go:39]: I don't want to change major parts of the reference implementation in this stage of the implementation, but I will keep this as a follow up task.

File: go/integration_test/processor/processor_gas_test.go
Comment (HerbertJordan) [reply go/integration_test/processor/processor_gas_test.go]: It might very likely be the case that in case of a failure the `GasUsed` field is ignored in Aida and the full gas consumption is just hard-coded.

It might also be that `receipts` are not checked at all in `aida-vm`.

**Review by HerbertJordan — APPROVED** [review]
Looks good, thanks!

**Pull request merged**
